libservo: Allow embedders to execute JavaScript scripts via the API

Narfinger opened the pull request:
This change adds a new `WebView` API `evaluate_javascript()`, which allows embedders to
execute JavaScript code and wait for a reply asynchronously. Ongoing script execution is
tracked by a libservo `JavaScriptEvaluator` struct, which maps an id to the callback passed
to the `evaluate_javascript()` method. The id is used to track the script and its execution
through the other parts of Servo.

Testing: This changes includes `WebView` unit tests.

Narfinger committed:
Allow embedders to execute js code in the current webview and recieve
the result

Webdriver now uses ConstellationMsg::EvaluateJavaScript for the async
version of WebdriverScriptCommand::ExecuteAsyncScript

Signed-off-by: Narfinger <Narfinger@users.noreply.github.com>

mrobinson commented on components/servo/webview.rs:
```diff
@@ -445,4 +446,17 @@ impl WebView {
         self.inner().compositor.borrow_mut().render()
     }
+
+    pub fn execute_js(&self, script: String, sender: IpcSender<WebDriverJSResult>) {
+        let bcid: BrowsingContextId = self.id().into();
+
+        self.inner()
+            .constellation_proxy
+            .send(ConstellationMsg::WebDriverCommand(
+                WebDriverCommandMsg::ScriptCommand(
+                    bcid,
+                    WebDriverScriptCommand::ExecuteScript(script, sender),
+                ),
+            ));
+    }
```
We are not going to be exposing `IpcSender` to the API. Let's have a conversation about how this API should look.

Narfinger replied:
Ok what would be a good way to do this then? I am not familiar yet with what should and shouldn't be done in servo but I didn't want to block in the webview impl.

jschwe commented on components/constellation/constellation.rs:
```diff
@@ -528,6 +533,25 @@ pub struct InitialConstellationState {
+struct EvaluateJavaScriptData {
+    /// Embedders can run arbitrary javascript via messages to the constellation.
+    /// We forward these messages to the script thread but need to store the callback function.
+    pub(crate) callbacks: HashMap<embedder_traits::ScriptId, fn(Result<JSValue, JSValueError>)>,
```
instead of storing a function, it would probably be nicer to store a `dyn` trait implementation.

mrobinson committed:
General reworking of this change:

1. Have callback be called in the embedder thread rather than in the
   constellation thread.
2. There is now a JavaScriptEvaluator that tracks all ongoing
   evaluations in the embedder.
3. Intead of reproducing the conversion code to produce JSValue, just
   convert WebDriver script values to JSValue. Eventually this should
   probably work the opposite way.
4. Add tests.
5. Remove unrelated WebDriver changes. Eventually the WebDriver will
   use this new API, but not yet.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

Narfinger committed:
Fix Formatting

Signed-off-by: Narfinger <Narfinger@users.noreply.github.com>

jdm reviewed (approved):
This setup makes sense to me!
